### DISCUSSION TITLE
How do I create an animation?

### DISCUSSION THREAD

**Discussion by yutik-nn** [bjlittle/geovista#400] [General]
Hi @bjlittle, it's Yuta. We've met at the SciPy. 

I want to create a 3D animation of the SSH on Earth (globally, not just a little domain) to use in my final presentation for the internship. 
You kindly showed me how to achieve a static result. Not I am wondering how exactly you created an animation that you have in the example with WaveWatch3. 
I have gone through the example code, but it seems to be for 1 step at a time, and the data from the pantry don't even seem to have the time dimension. 
Simply running the example doesn't produce the animated result. How do I do it? 

Thank you!!!

**Comment by bjlittle** [comment]
Hey @yutik-nn !

Great to hear from you 😀

Thanks for making my SciPy. It was so cool and a total pleasure to hangout with you during the SciPy sprints and introduce you to geovista.

I've not shared any example code for creating realtime time-series animations simply because of the amount of data involved. Initially, I just wanted to showcase that it was possible. Nevertheless, I'd love to help you 👍

Is it possible for you to share your data with me somehow? Is it publicly available somewhere that I can download?

I can show you how to run a time-series animation in real-time within a QT window, or within a Jupyter notebook and use [ipywidgets](https://ipywidgets.readthedocs.io/en/stable/) to control the animation with sliders.  It's also easy to create GIFs using PyVista. It all depends on how you want to present the animation.

**Reply by yutik-nn** [reply]
Thank you Bill! Conviniently, some SSH data has just been uploaded to the cloud. 
So if you run the following (I am giving you a Jupyter Notebook version since it's my go-to way to code). 

`ds = xr.open_dataset(
    "https://ncsa.osn.xsede.org/Pangeo/pangeo-cmems-duacs",
    consolidated=True, 
    engine="zarr"
)`
`ds = 
ds
    .drop_vars(['crs','nv','lat_bnds','lon_bnds', 'adt','err','ugos','ugosa','vgos','vgosa'])
  .sel(
        time=slice('1993-01-01','1994-01-01')
    )
`

Time slicing is optional. Given you have xarray and dask the code above should just work. The variable is 'sla'.

**Reply by yutik-nn** [reply]
It's been an absolute please to hang out with you at SciPy. Your tool is amazing, I am telling everyone about it when I get back to Hawai'i. 
Wow! Running an animation in Jupyter and also creating a GIF sounds dreamy and exactly what I want to learn. 
So grateful for your help!
And such a quick response!

**Reply by bjlittle** [reply]
Awesome thanks!

I'll make time to have a play this weekend and get something to you early next week, and we can roll from there to get it exactly as you need 👍🍻

**Reply by yutik-nn** [reply]
I actually need to use it on Monday already.
In the meantime I have a couple of follow up questions (if I may): 
1. Is there a way to 'shine more light' ? The figure that comes out seems too dark. 
![screenshot (1)](https://github.com/bjlittle/geovista/assets/64946569/69aedfed-ab1f-44d0-8319-47864359b7da)
2. If I needed to add points? Is there a way to plot points (that are coming from an xarray dataset and have lat, lon coordinates associated with it?). If it's not a mesh (or is it still a mesh) plotter.what would it be? 
So that I can have SSH covering the globe plus some points (tide gauge locations along the East Coast).

Thank you!

**Reply by bjlittle** [reply]
Ahh okay, nothing like a challenge.

There are several useful tutorials on  lighting in the PyVista documentation, see [here](https://docs.pyvista.org/version/stable/search.html?q=light).

Perhaps camera headlights are what you need?

Try incorporating the following into your code:

```python
import geovista as gv
import pyvista as pv

plotter = gv.GeoPlotter(lighting="none")
light = pv.Light(light_type="headlight")
plotter.add_light(light)
...
```

**Reply by bjlittle** [reply]
@yutik-nn With regards to adding points, yes you can do this 👍

I'm actually in the process of making this easier and have a pull-request in the making, which would also transform any points to the target projection of the plotter scene, but it's not landed yet. Soz.

However, there is a new example on the `main` branch that you can see the pattern of what to do in the meantime. That will suffice for your needs for now.

See the  [geovista.examples.earthquakes](https://github.com/bjlittle/geovista/blob/main/src/geovista/examples/earthquakes.py) example. I think this is exactly what you need.

Essentially, provide your lon/lat points to the [geovista.common.to_cartesian](https://github.com/bjlittle/geovista/blob/main/src/geovista/common.py#L617) function, which will return the xyz cartesian equivalent of your lon/lat points.

Then use the `add_points` method on the plotter, providing the xyz points and their data values (if you have any to colour the points accordingly), to render points on the spherical mesh.

Does this help?

**Reply by yutik-nn** [reply]
It's quite unbelievable! Thank you Bill. Nothing like talking to the creator. 
I was creating a mesh with only points and the rest of them NaN when there is `.to_cartesian `function
<img width="716" alt="image" src="https://github.com/bjlittle/geovista/assets/64946569/5409fbb9-0d8f-4af5-879b-658d963d56e5">

**Reply by bjlittle** [reply]
> It's quite unbelievable! Thank you Bill. Nothing like talking to the creator. 
> I was creating a mesh with only points and the rest of them NaN when there is `.to_cartesian `function
> <img width="716" alt="image" src="https://github.com/bjlittle/geovista/assets/64946569/5409fbb9-0d8f-4af5-879b-658d963d56e5">
> 

I love it! Nice one 👌

Does the use of headlight lighting help?

**Reply by yutik-nn** [reply]
I was looking for 'light' options in the repository. It's not explicitly mentioned anywhere, is it?

**Reply by bjlittle** [reply]
@yutik-nn There's no additional light controls in GeoVista, as they're not necessary. PyVista provides everything we need.

So hit the PyVista Documentation, see https://docs.pyvista.org/version/stable/search.html?q=light

Perhaps experiment with the different lighting controls and options to get what you want.

**Reply by yutik-nn** [reply]
Yes, thank you! I just keep forgetting geovista is built off of PyVista 😅

**Reply by yutik-nn** [reply]
and no, actually 'headlight' doesn't do much and 'three lights' does too much. 
Needs more experimenting ...

**Reply by bjlittle** [reply]
@yutik-nn Hey!

I managed to grab some spare time... and I've managed to easily get an ipywidget slider working within a Jupyter notebook to control the time dimension, so that you can slide through the time-series and render the scene. It's fairly responsive, which is nice. It's not as smooth as an animation, but it gives you the interactive control that you might desire.

I've been experimenting with the SMC time-series data, simply as I have it to hand. But once I'm happy with the pattern, then I'll use your data and send you a gist of my notebook.

This that okay?

**Reply by yutik-nn** [reply]
Hi @bjlittle ! I've tried but could not make things work. That's amazing! Thank you so very much. 
I am also slightly confused if `off_screen `is the key to subsequent saving of a movie 🤷🏼‍♀️

**Reply by bjlittle** [reply]
In this example I'm using an `IntSlider` with `continuous_update=True`, which will attempt to render each time-step when you move the slider. My laptop is a tad old, so it's not as smooth as it could be, but it's reasonable. Alternatively, setting `continuous_update=False` will only render the time-step that the slider stops on... I think you get the idea.

https://github.com/bjlittle/geovista/assets/2051656/f56dd817-1cd5-48c4-bc42-ae38fde2eed9

**Reply by bjlittle** [reply]
I'll try this now with your data and let you know how I get on :+1:

**Reply by bjlittle** [reply]
Okay, so I seem to be able to render your data okay...
![screenshot](https://github.com/bjlittle/geovista/assets/2051656/50521a88-bc4f-4fb9-b416-c8f9fbeef68c)

Note that you have the grid bounds data, which I've used instead of the grid cell centers

**Reply by yutik-nn** [reply]
When I try to use the code from your previous example in this data, widgets is not recognized. Do I need to import it in a special way?

**Reply by yutik-nn** [reply]
Oh I guess it's `import ipywidgets as widgets`

**Reply by yutik-nn** [reply]
I've been trying to experiment with [this](https://github.com/pyvista/pyvista-xarray/blob/main/examples/level_of_detail.py) example from pyvista-xarray but it doesn't even run for some reason. 
When trying to experiment with the line of code from your little video above it was asking what 'callback'. 
There is something fundamental that I do not understand how elements of the visualizations create and animation 😣

**Reply by bjlittle** [reply]
@yutik-nn Okay, try this [gist notebook](https://gist.github.com/bjlittle/e4f5ad4ad49834ed9d2a4dd88dc39e5b).

It's not perfect, but the best I can do at the moment.

Ideally, you'd be rendering on the server side where the data resides then streaming the render to the client browser. But that's not easily possible. So as a compromise, I've created a sub-sample of the dataset over `time` for the `sla` variable and save it locally. Then reload the dataset and render each slice controlled by a slider widget. This approach is more responsive as it's not pulling data from the server for each frame to process and render.

Also, to keep things reasonably performant, and simple, I've not performed a threshold of the data.

Let me know how that goes :+1:

**Reply by yutik-nn** [reply]
Right now it isn't working. I see continents and coastlines, but no data! For some reason and all of a sudden hvplot broke also. Where I tried to look at the data the usual way and see if it works ...

**Reply by yutik-nn** [reply]
Basically when I am running it with the data from the cloud, I get no data on the sphere. 
When I am using my local file I am getting the following error when trying to create a mesh:
`ValueError: Require a 1-D '(N+1,)' x-axis array, or 2-D '(N, 2)' x-axis array, got 0-D '()'.`

**Reply by yutik-nn** [reply]
Ok. I think I am getting there. Cloud data seem to be the problem. I am able to replicate what you wrote in the Yuta notebook for a smaller domain on the globe. Now, I am trying to make it so that it goes through time steps on its own and then saving. But cannot understand how that works. I am using a [link](https://docs.pyvista.org/version/stable/examples/02-plot/gif.html) Martin shared below but I do not get it. It uses wave.gif, that already exists. 
Thank you Bill for answering all my questions 🙏🏼
In your original example on the front page, the waves move without you touching a slider.

**Reply by bjlittle** [reply]
@yutik-nn No problem, glad to help where I can ...

I'll make time to share the code that'll generate an equivalent real-time animation for your data.

**Reply by bjlittle** [reply]
@yutik-nn How about this? :wink: 

https://github.com/bjlittle/geovista/assets/2051656/7893cf20-bdea-4ec6-9471-3b7082e346d1

**Reply by yutik-nn** [reply]
Wow @bjlittle ! This looks amazing.  For some reason I am not able to render my own data. Even when it's not animated the result comes out gray. I have got to figure it out. 
Can you please share the code for the above animation? So cool.

**Reply by trexfeathers** [reply]
Hi @yutik-nn, there might be some delay until @bjlittle comes back from holiday.

**Reply by bjlittle** [reply]
Thanks @trexfeathers 🥰

@yutik-nn and I caught up on slack 👍

**Reply by rajivjha0599** [reply]
hey @bjlittle great work !
Can you share me the code for same as i'm working on something similar.

**Comment by trexfeathers** [comment]
You could take a look at #330, which includes animation. I found this PyVista documentation very useful: [Create a GIF Movie — PyVista 0.41.1 documentation](https://docs.pyvista.org/version/stable/examples/02-plot/gif.html)

**Reply by yutik-nn** [reply]
Thank you, I will go read it.

**Reply by bjlittle** [reply]
@all-contributors please add @trexfeathers for question

**Reply by bjlittle** [reply]
Hmmm interesting the `@all-contributors` 🤖 doesn't respond to GH Discussion comments ... good to know 🤔
